peep.c

```diff
@@ -3868,6 +3868,98 @@ Perl_rpeep(pTHX_ OP *o)
             }
             break;
 
+        case OP_SUBSTR: {
+            OP *expr, *offs, *len;
+            /* Specialize substr($x, 0, $y) and substr($x,0,$y,"") */
+            /* Does this substr have 3-4 args and amiable flags? */
+            if (
+                (((o->op_private & 7) == 4) || ((o->op_private & 7) == 3))
```

Comment:
This looks correct, but it seems a bit magical.

At the very least it should use `OPpARG3_MASK`, but even better would be a `MAXARG3`.

Though the existing `pp_substr` code is full of that magic number.😒

Reply:
Always the trouble when copying existing bad code - do you go back and tidy up the source as well?

Reply:
There was one other strangeness that struck me in the copied code:
```
const int len_is_uv = len_iv ? SvIOK_UV(PL_stack_sp[0]) : 1;
```
at which point I cried and went on with my life.

Looking at it now I think it is clearer as:
```
const bool len_is_uv = len_iv == 0 || SvIOK_UV(PL_stack_sp[0]);
```
But the len_iv check was introduced in 83f78d1a27 which was to improve the behaviour of lvalue substr which `substr_chop` doesn't handle.

Reply:
Ok, I'll add a `MAXARG3` and make the changes.